third_party/nvfuser/csrc/lower_unroll.cpp

```diff
@@ -172,15 +177,16 @@ void UnrollPass::handle(Expr* expr) {
       // reduces to a scalar (issue #491)
       kir::ExprMutator::registerReplace(expr, inline_ite, nullptr);
     } else {
-      kir::ExprMutator::registerReplace(
-          expr, inline_ite, &for_loops_.back()->body());
+      kir::ExprMutator::registerReplace(expr, inline_ite);
     }
     if (expr != expr_with_predicate) {
       GpuLower::current()->propagateExprInfo(expr, expr_with_predicate);
     }
     inline_ite->thenBody().push_back(expr_with_predicate);
   } else if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
     handle(for_loop);
+  } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
```

Comment:
I'm guessing you're adding something in the original PR that introduces `IfThenElse` before the unroll pass. What is it?

Reply:
Yes, it is loop rotation (which happens right before the unroll pass):
```python
for i in range(n):
  statement1(i)
  statement2(i)
  statement3(i)
  statement4(i)
```
transform to
```python
if 0 < n:
  for i = 0:
    statement1(i)
    statement2(i)
for i ...:
  statement3(i)
  statement4(i)
  if i + 1 < n:
    statement1(i)
    statement2(i)
```
I am actually not materializing these conditions (because I think existing predicates should already cover all illegal access), so I am just generating:
```python
for i = 0:
  statement1(i)
  statement2(i)
for i ...:
  statement3(i)
  statement4(i)
  if true:
    statement1(i)
    statement2(i)
```
But the `if true` is still necessary because I use it as a special container to mark which part of the for loop is rotated from the next iteration.

Reply:
OK. Not about this PR, but another thing we should clean up is to make the pass dependencies more explicit. I suspect there's a pass that assumes there's no `kir::IfThenElse` in the incoming expr list, but I don't remember that's validated.

Reply:
> OK. Not about this PR, but another thing we should clean up is to make the pass dependencies more explicit. I suspect there's a pass that assumes there's no `kir::IfThenElse` in the incoming expr list, but I don't remember that's validated.

I think you are referring to the double buffer pass? Fortunately, there is a `TORCH_INTERNAL_ASSERT` on `handle(kir::IfThenElse* ite)`.

Reply:
Thanks, but it's not much about a specific pass. These assumptions on the lowering pass dependencies should be explicitly represented and enforced.

Reply:
That sound like something like a pass manager? Passes should use a unified data structure and should have some metadata stored in that data structure so that the pass manager can parse and decide the order of passes?